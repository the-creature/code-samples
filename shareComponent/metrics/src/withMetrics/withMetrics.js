import { withProps, compose, setPropTypes, lifecycle } from 'recompose';
import { exposeMetrics } from 'react-metrics';

/**
 * Gets the data needed for the dispatch function used in the lifecycle.
 * Relies on the propTypes to determine what data to look for in props
 * @module withMetrics
 * @param {Object} props - the current props for this mapper
 * @param {Object} types - the metrics propTypes to determine which data to deconstruct from props
 * @param {Function} dispatch - name of the metrics function to call
 * @returns {HOC} - {{metricsProps: {}, dispatch: *}}
 */
export const propsMapper = (props, types, dispatch) => {
  types = !!types ? types : {};
  const keys = Object.keys(types);
  const metricsProps = {};
  keys.map(key => Object.assign(metricsProps, { [key]: props[key] }));
  return {
    metricsProps,
    dispatch
  };
};

export const lifeCycleSpec = lifecycle({
  componentDidMount() {
    const { metrics, metricsProps, dispatch } = this.props;
    if (metricsProps && dispatch) {
      metrics[dispatch]([metricsProps]);
    }
  }
});

const withMetrics = (propTypes, dispatch) => {
  return compose(
    exposeMetrics,
    setPropTypes(propTypes),
    withProps(props => propsMapper(props, propTypes, dispatch)),
    lifeCycleSpec
  );
};

export default withMetrics;
