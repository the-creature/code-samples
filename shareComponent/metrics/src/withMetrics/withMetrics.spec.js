import React from 'react';
import { propsMapper, lifeCycleSpec } from './withMetrics';
import PropTypes from 'prop-types';

describe('withMetrics', () => {
  test('propsMapper should provide metricsProps', () => {
    const propTypes = {
      arrivalTime: PropTypes.string,
      isUpsell: PropTypes.bool
    };
    const props = {
      foo: 'bar',
      arrivalTime: '16:00',
      isUpsell: true
    };
    expect(propsMapper(props, propTypes, 'trackThePage')).toMatchSnapshot();
  });

  test('componentDidMount lifeCycleSpec with dispatch', () => {
    const props = {
      metrics: { trackThePage: jest.fn() },
      metricsProps: { foo: 'bar' },
      dispatch: 'trackThePage'
    };

    const BaseComponent = () => <div>Base Component</div>;
    BaseComponent.displayName = 'TestBaseComponent';

    const EnhancedComponent = lifeCycleSpec(BaseComponent);
    mount(<EnhancedComponent {...props} />);

    expect(props.metrics.trackThePage).toHaveBeenCalledWith([props.metricsProps]);
  });

  test('componentDidMount lifeCycleSpec without dispatch', () => {
    const props = {
      metrics: { trackThePage: jest.fn() }
    };

    const BaseComponent = () => <div>Base Component</div>;
    BaseComponent.displayName = 'TestBaseComponent';

    const EnhancedComponent = lifeCycleSpec(BaseComponent);
    mount(<EnhancedComponent {...props} />);

    expect(props.metrics.trackThePage).not.toHaveBeenCalled();
  });
});
