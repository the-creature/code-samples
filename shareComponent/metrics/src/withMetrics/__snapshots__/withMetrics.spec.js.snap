// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withMetrics propsMapper should provide metricsProps 1`] = `
Object {
  "dispatch": "trackThePage",
  "metricsProps": Object {
    "arrivalTime": "16:00",
    "isUpsell": true,
  },
}
`;
